doc/tools/coqrst/coqdomain.py

```diff
@@ -210,6 +214,15 @@ def _add_target(self, signode, name):
             signode['names'].append(name)
             signode['first'] = (not self.names)
             self._record_name(name, targetid, signode)
+        else:
```

Comment:
Can you help me understand why this is not always a duplicate? (given that the target_id is already known?)

Reply:
Not something I needed to understand to detect duplicates.

Sphinx calls `add_target_and_index`,
which calls `_add_target`,
which calls `_target_id`,
which doesn't add the id to `self.state.document.ids`,
so it wouldn't be a duplicate.

However, I don't what actually adds the target_id to `self.state.document.ids`.

The added dup message is independent from adding the local `make_id`; I had tried that before in #12182.

Reply:
I guess I'm still confused by the logic.  The code in the else branch duplicates the code in _record_name, which already checks for duplicates, so I wonder if it's just that maybe we should remove `if targetid not in self.state.document.ids:`.

Reply:
I have frequently been confused by the `coqdomain` logic.   It usually takes hours to understand what's going on and to make simple changes.  Do you want me to remove these lines?  I added them because you suggested it.  Is it really necessary to reengineer this right now?

Reply:
> I have frequently been confused by the coqdomain logic.

I agree it's complicated — Sphinx APIs are pretty complex.

>  Is it really necessary to reengineer them right now?

No, it's just that the more stuff we add and the more complicated it gets, I guess.  So every time we do I try to make sure that we don't add more complexity that we don't understand.

That code is fine by me.

Reply:
@cpitclaudel I tend to unmark as "resolved" the discussion that are still going on (even if the actual problem is fixed). For people following with GitHub web notifications (as opposed to e-mail notifications), it can be a bit difficult to notice those masked comments otherwise.

Reply:
> I wonder if it's just that maybe we should remove `if targetid not in self.state.document.ids:`.

I think it's there because the mapping from names to the target-id was not 1-1.  It was, in fact, generating duplicate target ids.  If the `if` isn't present, they would generate warnings, causing CI failures.  For all the time I've looked at this code, it didn't register that it was generating duplicate target ids until now.